### PR TITLE
autoscaling_instance_refresh - prepare modules for promotion

### LINKED ISSUE #2016: Add support for "skip_matching" preference in autoscaling_instance_refresh
### Summary

AWS enables users to turn on a "skip_matching" preference when starting an Instance Refresh. The Ansible module OTOH does not expose that preference.

### Issue Type

Feature Idea

### Component Name

autoscaling_instance_refresh

### Additional Information

Once implemented the following should work:
```yaml (paste below)
    - name: START INSTANCE REFRESH
      community.aws.ec2_asg_instance_refresh:
        name: "{{ asg_name }}"
        state: started
        preferences:
          skip_matching: true
```

Looking at the code of the current Ansible module, the only real change is to add it to the list of allowed preferences and document it. The plumbing is already there to support passing that preference to the AWS API.

### Code of Conduct

- [X] I agree to follow the Ansible Code of Conduct

### LINKED ISSUE #2019: Typo in autoscaling_instance_refresh module in check mode
### Summary

Typo on line 222 of `autoscaling_instance_refresh.py` , currently reading:
```
                    module.exit_json(chaned=False, msg="In check_mode - No active referesh found, nothing to cancel.")
```
`chaned` needs to get `changed`

### Issue Type

Bug Report

### Component Name

autoscaling_instance_refresh

### Ansible Version

N/A

### Collection Versions

N/A

### AWS SDK versions

N/A

### Configuration

N/A

### OS / Environment

_No response_

### Steps to Reproduce

N/A

### Expected Results

N/A

### Actual Results

N/A

### Code of Conduct

- [X] I agree to follow the Ansible Code of Conduct

### LINKED ISSUE #2120: autoscaling_instance_refresh: add "max_healthy_percentage" (MaxHealthyPercentage)
### Summary

I want to ensure that a new instance gets started first before an older one get killed. according to the [AWS documentation](https://docs.aws.amazon.com/cli/latest/reference/autoscaling/start-instance-refresh.html) this should be possible by setting `MaxHealthyPercentage` higher than 100.

### Issue Type

Feature Idea

### Component Name

autoscaling_instance_refresh

### Additional Information

```yaml (paste below)

- name: "Refresh: {{ aws_autoscaling_grpname }}"
  community.aws.autoscaling_instance_refresh:
    aws_access_key: "{{ vault_aws_access_key }}"
    aws_secret_key: "{{ vault_aws_secret_key }}"
    region: "{{ aws_region }}"
    name: "{{ aws_autoscaling_grpname }}"
    preferences:
      min_healthy_percentage: "{{ aws_autoscaling_min_healthy_perc }}"
      instance_warmup: "{{ aws_autoscaling_warmup }}"
      max_healthy_percentage: 110 # <------------------
    state: started
  register: refreshid
```

### Code of Conduct

- [X] I agree to follow the Ansible Code of Conduct

### PR DESCRIPTION
**Pull request by abikouo**
##### SUMMARY
Closes #2120 
Closes #2019 
Closes #2016 

Prepare modules ``autoscaling_instance_refresh`` and ``autoscaling_instance_refresh_info`` for promotion:
- Refactor modules to use common code from `ansible_collections.amazon.aws.plugins.module_utils.autoscaling`
- Add type hinting
- Update integration tests

##### ISSUE TYPE

- Feature Pull Request

### CODE REVIEW

File: plugins/modules/autoscaling_instance_refresh.py [hunk 1]
```diff
@@ -84,123 +98,143 @@
...
-    returned: success
-    type: int
-    sample: 5
+instance_refreshes:
+    description: Details of the instance refreshes for the Auto Scaling group..
+    returned: always
+    type: complex
+    contains:
+        instance_refresh_id:
+            description: instance refresh id
```
Comment (GomathiselviS) [inline plugins/modules/autoscaling_instance_refresh.py:111]:
```suggestion
            description: instance refresh id.
```

File: plugins/modules/autoscaling_instance_refresh.py [hunk 2]
```diff
@@ -84,123 +98,143 @@
...
+            returned: success
+            type: str
+            sample: "08b91cf7-8fa6-48af-b6a6-d227f40f1b9b"
+        auto_scaling_group_name:
+            description: Name of autoscaling group
```
Comment (GomathiselviS) [inline plugins/modules/autoscaling_instance_refresh.py:116]:
```suggestion
            description: Name of autoscaling group.
```

File: plugins/modules/autoscaling_instance_refresh.py [hunk 3]
```diff
@@ -84,123 +98,143 @@
...
+              - The following are the possible statuses
+              - Pending - The request was created, but the operation has not started.
+              - InProgress - The operation is in progress.
+              - Successful - The operation completed successfully.
+              - Failed - The operation failed to complete.
+                You can troubleshoot using the status reason and the scaling activities.
+              - Cancelling - An ongoing operation is being cancelled.
+                Cancellation does not roll back any replacements that have already been
+                completed, but it prevents new replacements from being started.
+              - Cancelled - The operation is cancelled.'
```
Comment (GomathiselviS) [inline plugins/modules/autoscaling_instance_refresh.py:132]:
```suggestion
              - Cancelled - The operation is cancelled.
```

File: plugins/modules/autoscaling_instance_refresh.py [hunk 4]
```diff
@@ -84,123 +98,143 @@
...
+            returned: success
+            type: str
+            sample: "2015-11-25T00:05:36.309Z"
+        end_time:
+            description: The date and time this ASG was created, in ISO 8601 format.
+            returned: success
+            type: str
+            sample: "2015-11-25T00:05:36.309Z"
+        percentage_complete:
+            description: the % of completeness
```
Comment (GomathiselviS) [inline plugins/modules/autoscaling_instance_refresh.py:164]:
```suggestion
            description: the % of completeness.
```

File: plugins/modules/autoscaling_instance_refresh.py [hunk 5]
```diff
@@ -84,123 +98,143 @@
...
+            returned: success
+            type: int
+            sample: 100
+        instances_to_update:
+            description: num. of instance to update
```
Comment (GomathiselviS) [inline plugins/modules/autoscaling_instance_refresh.py:169]:
```suggestion
            description: number of instances to update.
```

File: plugins/modules/autoscaling_instance_refresh.py [hunk 6]
```diff
@@ -84,123 +98,143 @@
...
 from ansible_collections.community.aws.plugins.module_utils.modules import AnsibleCommunityAWSModule as AnsibleAWSModule
 
 
-def start_or_cancel_instance_refresh(conn, module):
+def validate_healthy_percentage(preferences: Dict[str, Union[bool, int]]) -> Optional[str]:
+    min_healthy_percentage = preferences.get("min_healthy_percentage")
+    max_healthy_percentage = preferences.get("max_healthy_percentage")
+
+    if min_healthy_percentage is not None and (min_healthy_percentage < 0 or min_healthy_percentage > 90):
+        return "The value range for the min_healthy_percentage is 0 to 90."
```
Comment (GomathiselviS) [inline plugins/modules/autoscaling_instance_refresh.py:196]: Should the min_healthy_percentage value range between 0 and 90 or 0 and 100? I understand the default value is 90. However, what is the maximum allowable value for this parameter?

File: plugins/modules/autoscaling_instance_refresh.py [hunk 7]
```diff
@@ -57,6 +58,19 @@
...
+          - If enabled V(true), then Amazon EC2 Auto Scaling skips replacing instances that match the desired configuration.
+        type: bool
+      max_healthy_percentage:
+        description:
+          - Specifies the maximum percentage of the group that can be in service and healthy, or pending,
+            to support your workload when replacing instances.
+          - The value is expressed as a percentage of the desired capacity of the Auto Scaling group.
+          - Value range is V(100) to V(200).
+          - When specified, you must also specify O(preferences.min_healthy_percentage), and the difference between them cannot be greater than 100.
+        type: int
```
Comment (GomathiselviS) [inline plugins/modules/autoscaling_instance_refresh.py:73]: Should we include the version_added attribute for both of these parameters?

File: tests/integration/targets/autoscaling_instance_refresh/tasks/instance_refresh_info.yml [hunk 8]
```diff
@@ -0,0 +1,92 @@
+---
+- name: test getting info for an ASG name
```
Comment (GomathiselviS) [inline tests/integration/targets/autoscaling_instance_refresh/tasks/instance_refresh_info.yml:2]:
```suggestion
- name: Test getting info for an ASG name
```

Comment (GomathiselviS) [reply tests/integration/targets/autoscaling_instance_refresh/tasks/instance_refresh_info.yml:2]: I suggest that all tasks should have a specified name where the initial letter is capitalized. While we currently do not run ansible-lint on tests, integrating ansible-lint with tests in the future could potentially lead to numerous failures.

**Review by GomathiselviS — APPROVED** [review]

File: plugins/modules/autoscaling_instance_refresh.py
Comment (alinabuzachis) [inline plugins/modules/autoscaling_instance_refresh.py:1]: Can you please update  the documentation to use consistently M(...), O(...) and V(...)?

File: plugins/modules/autoscaling_instance_refresh.py [hunk 9]
```diff
@@ -84,123 +100,143 @@
...
-    returned: success
-    type: int
-    sample: 100
-instances_to_update:
-    description: num. of instance to update
-    returned: success
-    type: int
-    sample: 5
+instance_refreshes:
+    description: Details of the instance refreshes for the Auto Scaling group..
```
Comment (alinabuzachis) [inline plugins/modules/autoscaling_instance_refresh.py:108]:
```suggestion
    description: Details of the instance refreshes for the Auto Scaling group.
```

File: plugins/modules/autoscaling_instance_refresh.py [hunk 10]
```diff
@@ -84,123 +100,143 @@
...
+    returned: always
+    type: complex
+    contains:
+        instance_refresh_id:
+            description: instance refresh id.
```
Comment (alinabuzachis) [inline plugins/modules/autoscaling_instance_refresh.py:113]:
```suggestion
            description: Instance refresh id.
```

File: plugins/modules/autoscaling_instance_refresh_info.py
Comment (alinabuzachis) [inline plugins/modules/autoscaling_instance_refresh_info.py:1]: Can you please update the documentation to use consistently M(...), O(...) and V(...)?

File: tests/integration/targets/autoscaling_instance_refresh/tasks/instance_refresh_info.yml [hunk 11]
```diff
@@ -0,0 +1,101 @@
+---
+- name: Test getting info for an ASG name
+  autoscaling_instance_refresh_info:
+    name: "{{ asg_name }}"
+  register: output
+
+- name: Assert that the correct number of records are returned
+  assert:
+    that:
+      - output | community.general.json_query(inst_refresh_id_json_query) | unique | length == 7
```
Comment (alinabuzachis) [inline tests/integration/targets/autoscaling_instance_refresh/tasks/instance_refresh_info.yml:10]: Can you please replace `community.general.json_query` and use the json filter instead? We may want to limit dependencies on other collections.

**Review by alinabuzachis — APPROVED** [review]

**Pull request merged**